_posts/2020-04-07-how-to-start-with-opentelemetry-in-nodejs.markdown

```diff
@@ -0,0 +1,300 @@
+---
+img: /img/logo/otel-black-stacked.svg
+layout: post
+title: "How to start tracing with OpenTelemetry in NodeJS?"
+date: 2020-04-07 09:08:27
+categories: [post]
+tags: [nodejs, codeinstrumentation, opentelemetry, tracing]
+summary: "I developed an eight hours workshop about application monitoring and
+code instrumentation two years ago. This year I updated it to use OpenTelemetry
+and that's what I learned to instrument a NodeJS application."
+changefreq: daily
+---
+
+This post is to celebrate the first beta release for the OpenTelemetry NodeJS
+application <i class="fas fa-glass-cheers"></i>
+
+Recently I developed a workshop about code instrumentation and application
+monitoring. It is an 8 hours full immersion on logs, metrics, tracing and so on.
+I developed it last year and I gave it twice. Let me know if you are looking for
+something like that.
+
+Almost all of it is opensource but I didn't figure out a good way to make it
+usable without my brain for now. This year I updated it to use OpenTelemetry and
+InfluxDB v2.
+
+Anyway the application is called
+[ShopMany](https://github.com/gianarb/shopmany). This application does
+not return any useful information about its state. It is an e-commerce made of a
+bunch of services in various languages. Obviously one of them is in NodeJS and
+that's the one I am gonna show you today.
+
+**Discaimer**: I can not define myself as a NodeJS developer. I wrote a bunch of
+AngularJS single page application back in the day, I wrote some Cordova mobile
+applications ages ago. I am not writing any JS production code since 2015 more
+or less.
+
+## First approach
+
+I concluded the application instrumentation it was the day the maintainers
+tagged the first beta release. Overnight I had to update libraries and test
+code. Very luckily.
+
+The way I learned about how to properly instrument
+[discount](https://github.com/gianarb/shopmany/tree/master/discount) required a
+lot of digging in the actual
+[opentelemery-js](https://github.com/open-telemetry/opentelemetry-js) but
+luckily for us it has a lot of examples and the library is designed to load a
+bunch of useful modules that are able to instrument the application by itself.
+The community is very helpful and you can chat via
+[Gitter](http://gitter.im/open-telemetry/opentelemetry-js).
+
+## Getting Started
+
+I am using ExpressJS and OpenTelemetry has a plugin for it that you can load,
+and it instruments the app by itself, same for MongoDB that is the packaged I am
+using.
+
+Those are the dependencies I installed in my applications, all of them are
+provided by the repository I linked above:
+
+```
+"@opentelemetry/api": "^0.5.0",
+"@opentelemetry/exporter-jaeger": "^0.5.0",
+"@opentelemetry/node": "^0.5.0",
+"@opentelemetry/plugin-http": "^0.5.0",
+"@opentelemetry/plugin-dns": "^0.5.0",
```

Comment:
There is a problem with the dns plugin (see https://github.com/open-telemetry/opentelemetry-js/issues/612), i wouldn't recommend to use it